src/spark_rapids_ml/knn.py

```diff
@@ -0,0 +1,344 @@
+    alias,
+)
+from spark_rapids_ml.params import _CumlClass, _CumlParams
+
+
+class NearestNeighborsClass(_CumlClass):
+    @classmethod
+    def _cuml_cls(cls) -> List[type]:
+        from cuml import NearestNeighbors as cumlNearestNeighbors
+
+        return [cumlNearestNeighbors]
+
+    @classmethod
+    def _param_mapping(cls) -> Dict[str, Optional[str]]:
+        return {}
+
+    @classmethod
+    def _param_excludes(cls) -> List[str]:
+        return [
+            "handle",
+            "algorithm",
+            "metric",
+            "p",
+            "algo_params",
+            "metric_expanded",
+            "metric_params",
+            "output_type",
+        ]
+
+
+class _NearestNeighborsCumlParams(_CumlParams, HasInputCol, HasLabelCol):
+    """
+    Shared Spark Params for NearestNeighbor and NearestNeighborModel.
+    """
+
+    def setInputCol(self, value: str) -> "_NearestNeighborsCumlParams":
+        """
+        Sets the value of `inputCol`.
+        """
+        self.set_params(inputCol=value)
+        return self
+
+
+class _CumlEstimatorNearestNeighbors(_CumlEstimatorSupervised):
+    """
+    Base class for Cuml Nearest Neighbor.
+    """
+
+    def _pre_process_data(
+        self, dataset: DataFrame
+    ) -> Tuple[
+        List[Column], Optional[List[str]], int, Union[Type[FloatType], Type[DoubleType]]
+    ]:
+        (
+            select_cols,
+            multi_col_names,
+            dimension,
+            feature_type,
+        ) = super()._pre_process_data(dataset)
+
+        select_cols.append(col(alias.row_number))
+
+        return select_cols, multi_col_names, dimension, feature_type
+
+
+class NearestNeighbors(
+    NearestNeighborsClass, _CumlEstimatorNearestNeighbors, _NearestNeighborsCumlParams
+):
+    """
+    Examples
+    --------
+    >>> from sparkcuml.neighbors import NearestNeighbors
+    >>> data = [([1.0, 1.0],),
+    ...         ([2.0, 2.0],),
+    ...         ([3.0, 3.0],),]
+    >>> topk = 2
+    >>> gpu_knn = NearestNeighbors().setInputCol("features").setK(topk)
+    >>> data_df = spark.SparkContext.parallelize(data).map(lambda row: (row,)).toDF(["features"])
+    >>> gpu_knn.fit(data_df)
+    >>> query = [[1.0, 1.0], [3.0, 3.0]]
+    >>> query_df = spark.SparkContext.parallelize(query).map(lambda row: (row,)).toDF(["features"])
+    >>> distances_df, indices_df = gpu_knn.kneighbors(query_df, return_distance=True)
+    >>> distances_df.show()
+    [0, 1.414]
+    [0, 1.414]
+    >>> indices_df.show()
+    [0, 1]
+    [2, 1]
+    """
+
+    def __init__(self, **kwargs: Any) -> None:
+        super().__init__()
+        self.set_params(**kwargs)
+        self.label_isdata = 0
+        self.label_isquery = 1
+        self.row_number_col = alias.row_number
+        self.set_params(labelCol=alias.label)
+
+    def setK(self, value: int) -> "NearestNeighbors":
+        """
+        Sets the value of `k`.
+        """
+        self.set_params(n_neighbors=value)
+        return self
```

Comment:
This should be handled via parameter mapping, since mapping a Spark Param value to a differently named cuML param leads to weird behavior/inconsistencies.  So, add a `k` Spark ML Param and a `_param_mapping` entry to map Spark `k` to cuML `n_neighbors`.  (See `KMeans` for reference).

Reply:
done.